clientlibs/js/src/UpGradeClient/UpgradeClient.ts

```diff
@@ -0,0 +1,481 @@
+import { UpGradeClientInterfaces } from '../types';
+import {
+  IFeatureFlag,
+  ISingleMetric,
+  IGroupMetric,
+  ILogInput,
+  CaliperEnvelope,
+  IExperimentAssignmentv5,
+  MARKED_DECISION_POINT_STATUS,
+} from 'upgrade_types';
+import Assignment from '../Assignment/Assignment';
+import ApiService from '../ApiService/ApiService';
+import { DataService } from '../DataService/DataService';
+import { IConfigOptions } from './UpGradeClient.types';
+
+/**
+ * UpGradeClient is the main class for interacting with the UpGrade API.
+ *
+ * @example
+ * ```typescript
+ * import UpgradeClient from 'upgrade_client_lib/dist/browser';
+ * ```
+ *
+ * ```typescript
+ * import UpgradeClient from 'upgrade_client_lib/dist/node';
+ * ```
+ *
+ * General UpGrade types can also be accessed as named exports:
+ * ```typescript
+ * import UpgradeClient, { IExperimentAssignment } from 'upgrade_client_lib/dist/browser';
+ * ```
+ *
+ * SDK-Specific types can be accessed also:
+ * ```typescript
+ * import { UpGradeClientInterfaces } from 'upgrade_client_lib/dist/clientlibs/js/src/identifiers';
+ *
+ * const initResponse: UpGradeClientInterfaces.IUser = await upgradeClient.init();
+ * ```
+ */
+
+export default class UpgradeClient {
+  private apiService: ApiService;
+  private dataService: DataService;
+
+  // allow MARKED_DECISION_POINT_STATUS to be exposed on the client a la UpgradeClient.MARKED_DECISION_POINT_STATUS
+  // this will allow js users who are not using the upgrade types package to use this enum for markExperimentPoint()
+  public static MARKED_DECISION_POINT_STATUS = MARKED_DECISION_POINT_STATUS;
+
+  /**
+   * When constructing UpgradeClient, the user id, api host url, and "context" identifier are required.
+   * These will be attached to various API calls for this instance of the client.
+   *
+   * @example
+   *
+   * ```typescript
+   * // required
+   * const hostUrl: "htts://my-hosted-upgrade-api.com";
+   * const userId: "abc123";
+   * const context: "my-app-context-name";
+   *
+   * // not required, each is also optional
+   * const options: {
+   *   token: "someToken";
+   *   clientSessionId: "someSessionId";
+   * }
+   *
+   * const upgradeClient: UpgradeClient[] = new UpgradeClient(hostURL, userId, context);
+   * const upgradeClient: UpgradeClient[] = new UpgradeClient(hostURL, userId, context, options);
+   * ```
+   */
+
+  constructor(userId: string, hostUrl: string, context: string, options?: IConfigOptions) {
+    const config: UpGradeClientInterfaces.IConfig = {
+      apiVersion: 'v5',
```

Comment:
Instead of hardcoding you can use the major version from package.json. I like the way you used the variable defined during webpack build time.

Reply:
what would be the simplest way of doing that so that?

Reply:
It would be the same way as setting the value of USE_CUSTOM_HTTP_CLIENT and IS_BROWSER during the webpack build.
You will need to read the `version` from `package.json` and give it to the webpack.

Reply:
implemented, it is now parsing the version from package.json in the webpack file and injecting via a webpack variable